Covariant returns for RuntimeType

benaadams opened the pull request:
Unfortunately crashes during crossgen

benaadams committed:
Covariant returns for RuntimeType

MichalStrehovsky commented on src/mono/netcore/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -2454,12 +2454,20 @@ private RuntimeType[] GetNestedTypes_internal(string? displayName, BindingFlags
             }
         }
 
+        public static new RuntimeType GetTypeFromHandle(RuntimeTypeHandle handle)
```
Some of these are going to break analysis within IL Linker whenever they're encountered. IL Linker doesn't hardcode runtime implementation details such as these optimizations - we wouldn't want to hardcode it into linker. Linker operates on public APIs only.

E.g. this one is normally [an intrinsic](https://github.com/mono/linker/blob/01f8592b4dcbc5a5e5e395cc2dba31260e3d3114/src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs#L263-L264), so is e.g. `EmptyTypes`. This will make it more difficult for IL Linker to statically analyze what's going on in these spots.

Cc @vitek-karas

benaadams replied:
🤔 The covariant returns allow a lot of the casts back to `RuntimeType` to be dropped (haven't applied all the changes in dropping the casts); and also to devirtualized as direct calls as its a sealed type, rather than virtual calls via `Type`

MichalStrehovsky replied:
Do you have any perf numbers for this?

The cast should be pretty cheap because it's an exact match (the cast will trivially succeed in the part of the type check that RyuJIT typically inlines - it succeeds before actually calling into the casting logic).

benaadams replied:
Type checks then might not mean much?

![image](https://user-images.githubusercontent.com/1142958/100634781-9f0b6f80-3327-11eb-929a-b8033a645ba8.png)

Though it does also drop the virtual calls:

![image](https://user-images.githubusercontent.com/1142958/100635270-3d97d080-3328-11eb-8f4c-df5f03c39090.png)

MichalStrehovsky replied:
I meant the non-R2R case: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgZgARRwQA2cAJnlgEx4DCeA3gLABQe7l+WAjAGx14AsgAoA9sABWcAMYw8ogJSNWHVZQDseYbQWiA3Co4BfVkaA

RyuJIT generates two checks before calling the casting helper - is it a null object, or is it exactly the same type that we're casting to. If yes, it skips calling into runtime completely.

This change is still an improvement, but is it an improvement that would justify likely having to implement & maintain a IL linker feature? E.g. [just earlier today](https://github.com/mono/linker/pull/1655) I implemented intrinsic recognition of calls to `Type.UnderlyingSystemType` within linker motivated by the use within CoreLib - after this change, any call to get_UnderlyingSystemType on a RuntimeType would no longer be intrinsically recognized without linker work because this is a callvirt to a different method.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -769,12 +769,12 @@ public sealed override void ReorderArgumentArray(ref object?[] args, object stat
         // Return any exact bindings that may exist. (This method is not defined on the
         //  Binder and is used by RuntimeType.)
-        public static MethodBase? ExactBinding(MethodBase[] match, Type[] types, ParameterModifier[]? modifiers)
+        public static RuntimeConstructorInfo? ExactBinding(RuntimeConstructorInfo[] match, Type[] types, ParameterModifier[]? modifiers)
```
Using RuntimeConstructorInfo will make sharing of this code problematic with NativeAOT. 

DefaultBinder is a super slow path. An extra casts here won't make difference.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -2619,32 +2619,32 @@ public override EventInfo[] GetEvents(BindingFlags bindingAttr)
         }
 
-        public override Type[] GetInterfaces()
+        public override RuntimeType[] GetInterfaces()
```
Note that this is potential breaking change if the caller mutates the array and tries to store non-RuntimeType types into the array.

janvorli committed:
Fix problem with covariant returns with RuntimeType

Ben Adams has tried to add usage of covariant returns for the
RuntimeType as the overriding return type and it has uncovered an
ordering issue in the SystemDomain::LoadBaseSystemClasses w.r.t.

This change fixes it by shuffling parts of initialization around a bit.

ViktorHofer commented:
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

jeffhandley commented:
Hi @benaadams -- I wanted to see if this was still on your radar. There's some feedback remaining and a [change to our folder structure](https://github.com/dotnet/runtime/pull/45355#issuecomment-740665709) would also need to be accommodated. If you still want to move this forward, we can leave it open; otherwise we could close this PR--either is fine with us; just let us know. Thanks!

benaadams commented:
It interesting; but merge conflicts aside, sounds like there might be issues doing for runtime types specifically.

Also will cause additional merge conflicts for  #45121; might revisit another time
